File: java/rotateimage.java
public class Solution {
    
    private void rotatelayer(int[][] matrix, int x1, int x2){
        
    }
    
    public void rotate(int[][] matrix) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int m = matrix.length;
        if(m==0) return;
        int x1 = 0;
        int x2 = m-1;
        int tmp = 0;
        
        while(x1<x2){
            int n = x2-x1;
            //[x1,x1][x1,x2] -- [x1,x2][x2,x2]
            for(int i=0;i<n;i++){
                tmp = matrix[x1][x1+i];
                matrix[x1][x1+i] = matrix[x1+i][x2];
                matrix[x1+i][x2] = tmp;
            }
            //[x2,x2][x2,x1]
            for(int i=0;i<n;i++){
                tmp = matrix[x1][x1+i];
                matrix[x1][x1+i] = matrix[x2][x2-i];
                matrix[x2][x2-i] = tmp;
            }
            //[x2,x1][x1,x1]
            for(int i=0;i<n;i++){
                tmp = matrix[x1][x1+i];
                matrix[x1][x1+i] = matrix[x2-i][x1];
                matrix[x2-i][x1] = tmp;
            }
            x1++;
            x2--;
        }
    }
}

File: cpp/RotateImage.cpp
class Solution {
public:
    int tmp;
    int n;
    int start;
    int i;
    
    inline void swap(vector<vector<int> > &matrix, int i1, int j1, int i2, int j2) {
        tmp = matrix[i1][j1];
        matrix[i1][j1] = matrix[i2][j2];
        matrix[i2][j2] = tmp;
    }
    
    inline void swap(int& a, int&  b) {
        int tmp = a;
        a = b;
        b = tmp;
    }

    void rotate(vector<vector<int> > &matrix) {
        n = matrix.size();
        if(n<=1) {
            return;
        }
        for(start = 0; start < n-1-start; start++) {
            for(i=start;i<n-1-start;i++) {
                swap(matrix[start][i], matrix[i][n-1-start]);
            }
            for(i=start;i<n-1-start;i++) {
                swap(matrix[start][i], matrix[n-1-start][n-1-i]);
            }
            for(i=start;i<n-1-start;i++) {
                swap(matrix[start][i], matrix[n-1-i][start]);
            }
        }
    }
};
